src/main/java/com/xpn/xwiki/plugin/lucene/searcherProvider/SearcherProvider.java

```diff
@@ -38,82 +41,72 @@
    * List of Lucene indexes used for searching. By default there is only one such index
    * for all the wiki. One searches is created for each entry in {@link #indexDirs}.
    */
-  private Searcher[] searchers;
+  private final Searcher[] backedSearchers;
+
+  private volatile boolean markToClose;
 
-  private boolean markToClose;
+  private volatile boolean isClosed;
 
   private Set<Thread> connectedThreads;
 
-  private HashMap<Thread, Set<SearchResults>> connectedSearchResultsMap;
+  private ConcurrentMap<Thread, Set<SearchResults>> connectedSearchResultsMap;
 
   SearcherProvider(Searcher[] searchers) {
-    this.searchers = searchers;
+    this.backedSearchers = searchers;
     this.markToClose = false;
-    this.connectedThreads = new HashSet<Thread>();
-    this.connectedSearchResultsMap = new HashMap<Thread, Set<SearchResults>>();
+    this.connectedThreads = Collections.newSetFromMap(new ConcurrentHashMap<Thread, Boolean>());
+    this.connectedSearchResultsMap = new ConcurrentHashMap<Thread, Set<SearchResults>>();
     LOGGER.debug("create searcherProvider: [" + System.identityHashCode(this) + "].");
   }
 
-  HashMap<Thread, Set<SearchResults>> internal_getConnectedSearchResults() {
+  ConcurrentMap<Thread, Set<SearchResults>> internal_getConnectedSearchResults() {
     return connectedSearchResultsMap;
   }
 
   Set<Thread> internal_getConnectedThreads() {
     return connectedThreads;
   }
 
-  /**
-   * connect and markToClose need to be externally synchronized
-   */
   public void connect() {
     if (!checkConnected()) {
-      if (this.markToClose) {
-        throw new IllegalStateException(
-            "you may not connect to a SearchProvider" + " marked to close.");
-      }
-      LOGGER.debug("connect searcherProvider [" + System.identityHashCode(this) + "] to ["
-          + Thread.currentThread() + "].");
+      checkState(!markToClose, "you may not connect to a SearchProvider marked to close.");
+      LOGGER.debug("connect searcherProvider [{}] to [{}].", System.identityHashCode(this),
+          Thread.currentThread());
       connectedThreads.add(Thread.currentThread());
```

Comment:
`checkState(...)` and `connectedThreads.add(...)` should be in a synchronized block, otherwise a thread could still get connected after the provider was marked to close